reader/src/main/java/org/jline/reader/PrintAboveWriter.java

```diff
@@ -0,0 +1,58 @@
+package org.jline.reader;
+
+import java.io.IOException;
+import java.io.Writer;
+
+/**
+ * Redirects a {@link Writer} to a {@link LineReader}'s {@link LineReader#printAbove(String)} method,
+ * which draws output above the current prompt / input line.
+ *
+ * <p>Example:</p>
+ * <pre>
+ *     LineReader reader = LineReaderBuilder.builder().terminal(terminal).parser(parser).build();
+ *     PrintAboveWriter printAbove = new PrintAboveWriter(reader);
+ *     printAbove.write(new char[] { 'h', 'i', '!', '\n'});
+ * </pre>
+ *
+ */
+public class PrintAboveWriter extends Writer {
+    private LineReader reader;
+    private StringBuilder stringBuffer;
+
+    private static int defaultCharBufferSize = 8192;
+
+    public PrintAboveWriter(LineReader reader) { this(reader, defaultCharBufferSize); }
+
+    public PrintAboveWriter(LineReader reader, int bufferSize) {
+        if (bufferSize <= 0)
+            throw new IllegalArgumentException("Buffer size <= 0");
+        this.reader = reader;
+        stringBuffer = new StringBuilder(bufferSize);
+    }
+
+    @Override
+    public void write(char[] cbuf, int off, int len) throws IOException {
+        stringBuffer.append(cbuf, off, len);
+    }
+
+    private void ensureOpen() throws IOException {
+        if (reader == null || stringBuffer == null)
+            throw new IOException("Stream closed");
+    }
+
+    @Override
+    public void flush() throws IOException {
+        ensureOpen();
+        int lastNewline = stringBuffer.lastIndexOf("\n");
+        if (stringBuffer.length() > 0 && lastNewline != -1) {
```

Comment:
The `stringBuffer.length() > 0` test looks superfluous at first glance. A `lastNewLine >= 0` should be sufficient I think.

Reply:
Fixed.